### PR TITLE
Correct attention mask dtype for Flax GPT2

### LINKED ISSUE #25634: Problem caused by boolean attention mask in `pretrained_model.generate` of Flax GPT2
Hi! 

I notice that the usage of a boolean attention mask in `pretrained_model.generate` of Flax GPT2 can cause an error. Here is a short, self-contained code block to showcase the problem; I also prepared a [colab notebook here](https://colab.research.google.com/drive/1fIfOr0AFfWlAho1dwuk8zqxKxlKmzd7i?usp=sharing):

``` python
import transformers
import jax
import jax.numpy as jnp

tokenizer = transformers.AutoTokenizer.from_pretrained(
  "gpt2", padding_side="right")
tokenizer.pad_token = tokenizer.eos_token
query = jnp.array([
    [tokenizer.pad_token_id, tokenizer.pad_token_id, 23073],
])

response_length = 4
# temperature = 0.7
pretrained_model = transformers.FlaxAutoModelForCausalLM.from_pretrained("gpt2")
generation_config = transformers.GenerationConfig(
    max_new_tokens=response_length,
    min_new_tokens=response_length,
    do_sample=True,
)
generation_config.pad_token_id = tokenizer.pad_token_id

context_length = query.shape[1]
attention_mask = query != tokenizer.pad_token_id

input_ids = query.clone()

# set padding tokens to 0
input_ids = jnp.where(attention_mask, input_ids, 0)
output = pretrained_model.generate(
    input_ids=input_ids,
    attention_mask=attention_mask,
    generation_config=generation_config,
)

# TypeError: lax.dynamic_update_slice requires arguments to have the same dtypes, got int32, bool.
```

The type error occurs because the `attention_mask` in our example above is a boolean array. But the `extended_attention_mask` used in [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_flax_gpt2.py#L753) internally for response generation has an integer type. This leads to an error in the `lax.dynamic_update_slice` [line here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_flax_gpt2.py#L756), as it can't handle inputs with different data types (integer and boolean).

I think this can be a bug, because boolean attention mask should be permitted.
To fix it, one can simply update [this line](https://github.com/huggingface/transformers/blob/main/src/transformers/models/gpt2/modeling_flax_gpt2.py#L756) in `transformers.models.gpt2.modelling_flax_gpt2.py`, which currently reads
`extended_attention_mask = lax.dynamic_update_slice(extended_attention_mask, attention_mask, (0, 0))`
into the following new line:
`extended_attention_mask = lax.dynamic_update_slice(extended_attention_mask, attention_mask.astype("i4"), (0, 0))`
This will correct the mismatch in dtypes. 

Happy to submit a PR for that!

### Who can help?

@sanchit-gandhi, @gante

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

Here is a short, self-contained code block to showcase the problem; I also prepared a [colab notebook here](https://colab.research.google.com/drive/1fIfOr0AFfWlAho1dwuk8zqxKxlKmzd7i?usp=sharing):

``` python
import torch
import transformers
import jax
import jax.numpy as jnp

tokenizer = transformers.AutoTokenizer.from_pretrained(
  "gpt2", padding_side="right")
tokenizer.pad_token = tokenizer.eos_token
query = jnp.array([
    [tokenizer.pad_token_id, tokenizer.pad_token_id, 23073],
])

response_length = 4
# temperature = 0.7
pretrained_model = transformers.FlaxAutoModelForCausalLM.from_pretrained("gpt2")
generation_config = transformers.GenerationConfig(
    max_new_tokens=response_length,
    min_new_tokens=response_length,
    do_sample=True,
)
generation_config.pad_token_id = tokenizer.pad_token_id

context_length = query.shape[1]
attention_mask = query != tokenizer.pad_token_id

input_ids = query.clone()

# set padding tokens to 0
input_ids = jnp.where(attention_mask, input_ids, 0)
output = pretrained_model.generate(
    input_ids=input_ids,
    attention_mask=attention_mask,
    generation_config=generation_config,
)
# TypeError: lax.dynamic_update_slice requires arguments to have the same dtypes, got int32, bool.
```

### Expected behavior

I expected to execute the line `output = pretrained_model.generate(
    input_ids=input_ids,
    attention_mask=attention_mask,
    generation_config=generation_config,
)` in the above example, when `attention_mask` is a boolean mask.

### PR DESCRIPTION
**Pull request by liutianlin0121**
# What does this PR do?

Fixes #25634

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case. Link: https://github.com/huggingface/transformers/issues/25634
- [N/A] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sanchit-gandhi 
@ArthurZucker

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
Nice catch! Could you maybe add a test in the `test_modelling_flax_gpt2.py` to make sure this is tested? 😉 (taking inspiration from your minimal reproducer!)

**Comment by liutianlin0121** [comment]
@ArthurZucker Sure! I added a test :-)

File: tests/models/gpt2/test_modeling_flax_gpt2.py [hunk 1]
```diff
@@ -335,3 +335,42 @@ def test_model_from_pretrained(self):
             model = model_class_name.from_pretrained("gpt2", from_pt=True)
             outputs = model(np.ones((1, 1)))
             self.assertIsNotNone(outputs)
+
+    def test_bool_attention_mask_in_generation(self):
```
Comment (sanchit-gandhi) [inline tests/models/gpt2/test_modeling_flax_gpt2.py:339]: Could we make this a fast test? We can first define a `check_bool_attention_mask_in_generation ` function in the model tester:
```python
    def check_bool_attention_mask_in_generation(self, model_class_name, config, input_ids, attention_mask):
        model = model_class_name(config)
        attention_mask_bool = input_ids != 0

        output_bool_att_mask = model.generate(
            input_ids=input_ids,
            attention_mask=attention_mask_bool,
            max_new_tokens=3,
        )

        output_int_att_mask = model.generate(
            input_ids=input_ids,
            attention_mask=attention_mask_int,
            max_new_tokens=3,
        )

        self.assertListEqual(output_bool_att_mask.sequences,  output_int_att_mask.sequences, "Generated response differ between boolean and integer attention mask",)
```

And then a test function in the model test:
```python
    def test_bool_attention_mask_in_generation(self):
        for model_class_name in all_generative_model_classes:
             config, input_ids, attention_mask = self.model_tester.prepare_config_and_inputs()
            self.model_tester. check_bool_attention_mask_in_generation(model_class_name, config, input_ids, attention_mask)
```

**Review by sanchit-gandhi — COMMENTED** [review]
Modelling code changes LGTM - would be great to make the test a fast test by defining a tester function in the model tester, and then executing it in the model test

**Comment by liutianlin0121** [comment]
>would be great to make the test a fast test by defining a tester function in the model tester, and then executing it in the model test

@sanchit-gandhi Good point! Done.

**Review by ArthurZucker — APPROVED** [review]
Thanks for adding a test! 🤗

**Comment by ArthurZucker** [comment]
cc @sanchit-gandhi feel free to merge if it's alright with you!

File: tests/models/gpt2/test_modeling_flax_gpt2.py [hunk 2]
```diff
@@ -187,6 +187,27 @@ def check_use_cache_forward_with_attn_mask(self, model_class_name, config, input
...
+            input_ids=input_ids,
+            attention_mask=attention_mask.astype(bool),
+            max_new_tokens=3,
+        )
+
+        np.testing.assert_array_equal(
+            output_bool_att_mask.sequences,
+            output_int_att_mask.sequences,
+            "Generated response differ between boolean and integer attention mask",
+        )
```
[the remark this answers is not in the document]
Comment (sanchit-gandhi) [reply tests/models/gpt2/test_modeling_flax_gpt2.py]: Sorry should be `self.parent.assertTrue` - apologies!

**Comment by liutianlin0121** [comment]
@sanchit-gandhi Hey thanks! I change to assertTrue.

**Review by sanchit-gandhi — APPROVED** [review]
Awesome - thanks @liutianlin0121!

**Comment by liutianlin0121** [comment]
No problem! Feel free to merge it (it seems that I can't).

**Pull request merged**